guides/creating_a_new_component.md

```diff
@@ -0,0 +1,386 @@
+        """
+        return x
+```
+
+
+Once defined, it is necessary to import the new class inside the [\_\_init\_\_](https://github.com/gradio-app/gradio/blob/main/gradio/__init__.py) module class in order to make it module visible.
+
+```python
+
+from gradio.components import (
+    ...
+    ColorPicker,
+    ...
+)
+
+```
+
+### Step 1.1 - Writing Unit Test for Python Class
+
+When developing new components, you should also write a suite of unit tests for it. The tests should be placed in the [gradio/test/test_components.py](https://github.com/gradio-app/gradio/blob/main/test/test_components.py) file. Again, as above, take a cue from the tests of other components (e.g. [Textbox](https://github.com/gradio-app/gradio/blob/main/test/test_components.py)) and add as many unit tests as you think are appropriate to test all the different aspects and functionalities of the new component. For example, the following tests were added for the ColorPicker component:
+
+```python
+class TestColorPicker(unittest.TestCase):
+    def test_component_functions(self):
+        """
+        Preprocess, postprocess, serialize, save_flagged, restore_flagged, tokenize, generate_sample, get_config
+        """
+        color_picker_input = gr.ColorPicker()
+        self.assertEqual(color_picker_input.preprocess("#000000"), "#000000")
+        self.assertEqual(color_picker_input.preprocess_example("#000000"), "#000000")
+        self.assertEqual(color_picker_input.postprocess(None), None)
+        self.assertEqual(color_picker_input.postprocess("#FFFFFF"), "#FFFFFF")
+        self.assertEqual(color_picker_input.serialize("#000000", True), "#000000")
+
+        color_picker_input.interpretation_replacement = "unknown"
+
+        self.assertEqual(
+            color_picker_input.get_config(),
+            {
+                "value": None,
+                "show_label": True,
+                "label": None,
+                "style": {},
+                "elem_id": None,
+                "visible": True,
+                "interactive": None,
+                "name": "colorpicker",
+            },
+        )
+        self.assertIsInstance(color_picker_input.generate_sample(), str)
+
+    def test_in_interface_as_input(self):
+        """
+        Interface, process, interpret,
+        """
+        iface = gr.Interface(lambda x: x, "colorpicker", "colorpicker")
+        self.assertEqual(iface.process(["#000000"]), ["#000000"])
+
+    def test_in_interface_as_output(self):
+        """
+        Interface, process
+
+        """
+        iface = gr.Interface(lambda x: x, "colorpicker", gr.ColorPicker())
+        self.assertEqual(iface.process(["#000000"]), ["#000000"])
+
+    def test_static(self):
+        """
+        postprocess
+        """
+        component = gr.ColorPicker("#000000")
+        self.assertEqual(component.get_config().get("value"), "#000000")
+```
+
+## Step 2 - Create a New Svelte Component
+
```

Comment:
It would be good to add a comment here about why the components appear in two different places.

The 'core' component that lives in `packages/component-name` is the generic component that has no knowledge of gradio specific functionality. The component found in `packages/app/src/components` is a Gradio specific wrapper that handles Gradio specific logic and passes the necessary data down to the core component and attaches any necessary event listeners.

Reply:
Perfect, this explanation has been added to step 2. Thanks!!